<a href="https://colab.research.google.com/github/andrewbeeksma/eBay-Car-Sales/blob/master/eBay_Car_Sales.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# eBay Car Sales Analysis

In this project, I will be cleaning and analyzing a data set from Kaggle that contains information on used cars. I'll be practicing fundamental data cleaning techniques, as well as using the NumPy and pandas libraries to perform vectorized operations for analysis.

In [ ]:
import numpy as np
import pandas as pd
autos = pd.read_csv('autos.csv', encoding='Latin-1')

In [ ]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [ ]:
print(autos.info())
print(autos.head())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

There's a few things to note from the data set above. The first is that the data set contained 20 columns, most of which are strings. The second thing worth noting is that some of the columns have null values, but not this is a relatively small percentage. And finally, it is clear that column names use camelcase, rather than Python's traditional snakecase, so I'll be changing those.

In [ ]:
print(autos.columns)

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


In [ ]:
autos.rename(columns={'dateCrawled': 'date_crawled',
             'offerType': 'offer_type',
             'vehicleType': 'vehicle_type',
             'yearOfRegistration': 'registration_year',
             'monthOfRegistration': 'registration_month',
             'fuelType': 'fuel_type',
             'notRepairedDamage': 'unrepaired_damage',
             'dateCreated': 'ad_created',
             'nrOfPictures': 'nr_of_pictures',
             'postalCode': 'postal_code',
             'lastSeen': 'last_seen'}, inplace=True)
print(autos.head())

          date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  powerPS   model   odometer  registration_month fuel_type  \
0    manuell      158  a

In [ ]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-19 17:36:18  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox       powerPS  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

It is worth noting that the 'seller' and 'offer_type' columns have values that are nearly all identical. Note that out of 50,000 rows, each columns has the same data point for 49,999 of those rows. This will be useful to keep in mind later during my analysis. Another one of the things that is clear from the table above is that the 'price' and 'odometer' columns store their numeric data as texts. Consequently, I'll go through these columns cleaning up the values by removing non-numeric characters and converting the text to numeric data types.

In [ ]:
print(autos['price'].head(3))
print(autos['odometer'].head(3))

0    $5,000
1    $8,500
2    $8,990
Name: price, dtype: object
0    150,000km
1    150,000km
2     70,000km
Name: odometer, dtype: object


In [ ]:
autos['price'] = autos['price'].str.replace('$','').str.replace(',','').astype(int)
autos['odometer'] = autos['odometer'].str.replace(',','').str.replace('km','').astype(int)
autos.rename(columns={'odometer': 'odometer_km'}, inplace=True)

print(autos['price'].head(3))
print(autos['odometer_km'].head(3))

0    5000
1    8500
2    8990
Name: price, dtype: int64
0    150000
1    150000
2     70000
Name: odometer_km, dtype: int64


Next, I'll be exploring the 'price' and 'odometer_km' columns using the pandas library, determining if there are data points that are clearly incorrect and/or unrealistic. I can also determine the existence of outliers and deal with those accordingly.

In [ ]:
print(autos['price'].shape)
#print(autos['price'].describe())
print(autos['price'].value_counts().sort_index(ascending=False).head(20))
print(autos['price'].value_counts().sort_index(ascending=True).head(50))

print(autos['odometer_km'].shape)
#print(autos['price'].describe())
print(autos['odometer_km'].value_counts().sort_index(ascending=True))


(50000,)
99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price, dtype: int64
0      1421
1       156
2         3
3         1
5         2
8         1
9         1
10        7
11        2
12        3
13        2
14        1
15        2
17        3
18        1
20        4
25        5
29        1
30        7
35        1
40        6
45        4
47        1
49        4
50       49
55        2
59        1
60        9
65        5
66        1
70       10
75        5
79        1
80       15
89        1
90        5
99       19
100     134
110       3
111       2
115       2
117       1
120      39
122       1
125       8
129       1
130      15
135       1
139       1
140       9
Name: price, dtype: int64
(50000,)
5000        967
10000       264
20000       784
30

There are a few things to note about the data above. First, in the 'price' column, it is clear that there are numerous cars above the price of 999,999 dollars. Not only that, but there are cars listed for over the price of 10,000,000 dollars. Similarly, there are many cars that are listed to be less than 100 dollars, with over 1,000 listings for cars at a price of 0 dollars. Furthermore, in the 'odometer_km' column, most rows increment by 10,000 kilometers, but the final row lists 32,424 entries for cars with an odometer of 150,000 km. Because of this, we can assume that most of these cars are well over 150,000 km, and it might help to think of this row more as '150,000+'. So, the first thing I'm going to do next is take care of the outliers in the 'price' column.

In [ ]:
autos[(autos['price'] < 100) | (autos['price'] > 999999)] = autos['price'].between(100,999999).mean().astype(int)
print(autos['price'].value_counts().sort_index(ascending=True).head())
print(autos.describe(include='all'))

0      1773
100     134
110       3
111       2
115       2
Name: price, dtype: int64
        date_crawled     name  seller offer_type          price abtest  \
count        50000.0  50000.0   50000      50000   50000.000000  50000   
unique       46575.0  37165.0       3          2            NaN      3   
top              0.0      0.0  privat    Angebot            NaN   test   
freq          1773.0   1773.0   48226      48227            NaN  24829   
mean             NaN      NaN     NaN        NaN    5779.724400    NaN   
std              NaN      NaN     NaN        NaN   11832.368937    NaN   
min              NaN      NaN     NaN        NaN       0.000000    NaN   
25%              NaN      NaN     NaN        NaN    1100.000000    NaN   
50%              NaN      NaN     NaN        NaN    2950.000000    NaN   
75%              NaN      NaN     NaN        NaN    7200.000000    NaN   
max              NaN      NaN     NaN        NaN  999999.000000    NaN   

       vehicle_type  regi

The next thing I'm going to look at is the 'date' columns. Currently, there are five columns that contain data relevant to a particular date and time. These columns include 'date_crawled,' 'last_seen,' 'ad_created,' 'registration_month,' and 'registration_year.' What's important to understand about these columns is that the 'date_crawled,' 'last_seen,' and 'ad_created' columns are all strings, whereas I'm going to want to convert these to numeric data types, similar to 'registration_month' and 'registration_year.' Let's take a look at the string date columns.

In [ ]:
autos[['date_crawled','last_seen','ad_created']][0:5]

date_crawled            last_seen           ad_created
0  2016-03-26 17:47:46  2016-04-06 06:45:54  2016-03-26 00:00:00
1  2016-04-04 13:38:56  2016-04-06 14:45:08  2016-04-04 00:00:00
2  2016-03-26 18:57:24  2016-04-06 20:15:37  2016-03-26 00:00:00
3  2016-03-12 16:58:10  2016-03-15 03:16:28  2016-03-12 00:00:00
4  2016-04-01 14:38:50  2016-04-01 14:38:50  2016-04-01 00:00:00

From the table above, it is clear that the first 10 characters from each string represents the day. So, next I'm going to calculate the distribution of date values by day.

In [ ]:
print(autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=False).head())
print(autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=False).head())
print(autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=False).head())

2016-04-07    0.00134
2016-04-06    0.00306
2016-04-05    0.01260
2016-04-04    0.03524
2016-04-03    0.03724
Name: date_crawled, dtype: float64
2016-04-07    0.12746
2016-04-06    0.21410
2016-04-05    0.12062
2016-04-04    0.02366
2016-04-03    0.02424
Name: last_seen, dtype: float64
2016-04-07    0.00120
2016-04-06    0.00314
2016-04-05    0.01138
2016-04-04    0.03558
2016-04-03    0.03748
Name: ad_created, dtype: float64


In [ ]:
autos['registration_year'].describe()

count    50000.000000
mean      1933.642420
std        380.673798
min          0.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

Notice that the minimum date for the registration_year is 0, and the maximum date for the registration_year is 9999. Clearly both of these values are false, and it's also safe to assume that any registration_year after 2016, the year these listings were scraped off the web, is not an accurate data point. So first I'm going to see if it's safe to exclusively deal with data points that have registration years between 1900 and 2016.

In [ ]:
autos.drop(autos[~(autos['registration_year'].between(1900, 2016))].index, inplace=True)

In [ ]:
autos['registration_year'].value_counts().sort_index(ascending=True)

1910       2
1927       1
1929       1
1931       1
1934       2
        ... 
2012    1308
2013     801
2014     662
2015     380
2016    1202
Name: registration_year, Length: 78, dtype: int64

Next I'm going to perform aggregation to better understand the 'brand' column. This is a common data analysis technique and using these methods for this data set will help be get a better grasp of aggregating data using NumPy and pandas.

In [ ]:
autos['brand'].value_counts().head(10)

volkswagen       9801
bmw              5107
opel             4971
mercedes_benz    4480
audi             4022
ford             3238
renault          2182
peugeot          1384
fiat             1187
seat              846
Name: brand, dtype: int64

In [ ]:
avg_price_by_brand = {}
for brand in autos['brand'].value_counts().head(10).index:
    avg_price = autos[autos['brand'] == brand]['price'].mean()
    avg_price_by_brand[brand] = int(avg_price)
print(avg_price_by_brand)

{'volkswagen': 5639, 'bmw': 8381, 'opel': 3005, 'mercedes_benz': 8672, 'audi': 9380, 'ford': 4086, 'renault': 2496, 'peugeot': 3113, 'fiat': 2836, 'seat': 4433}


From this data, it becomes very clear that 'audi,' 'bmw,' and 'mercedes_benz' have the most expensive vehicles. In contrast, 'renault,' 'fiat,' 'opel,' and 'peugeot' have very cheap cars. And finally, in between we have 'ford,' 'volkswagen,' and 'seat.' Next, I'd also like to look at the average mileage for each of these brands, and try to find a correlation between the mileage and the brand name. This will help me to verify that the above data is correct, or possible undermine its validity.

In [ ]:
avg_mileage_by_brand = {}
for brand in autos['brand'].value_counts().head(10).index:
    avg_mileage = autos[autos['brand'] == brand]['odometer_km'].mean()
    avg_mileage_by_brand[brand] = int(avg_mileage)

apbb_series = pd.Series(avg_price_by_brand)
ambb_series = pd.Series(avg_mileage_by_brand)

price_to_mileage = pd.DataFrame(apbb_series, columns=['mean_price'])
price_to_mileage['mean_mileage'] = ambb_series
print(price_to_mileage)

               mean_price  mean_mileage
volkswagen           5639        128804
bmw                  8381        132695
opel                 3005        129384
mercedes_benz        8672        131025
audi                 9380        129245
ford                 4086        124277
renault              2496        128281
peugeot              3113        127127
fiat                 2836        116950
seat                 4433        121536


Awesome! I've populated a new pandas DataFrame using the dictionaries from above, which both assign average mileage and average price to a brand name. Based on the table above, it becomes increasingly clear that there is not a large disparity in the average mileage by brand name. Thus I can conclude that brand names such as bmw and mercedes_benz are not more expensive simply because they tend to have fewer miles on them. Rather, they go for higher because of the higher quality nature of the vehicle.